File: 1391-check-if-there-is-a-valid-path-in-a-grid/solution.cpp
#include <vector>
#include <queue>
#include <map>

using namespace std;

class Solution {
    // Mapping each street type to the directions it connects
    // Directions: {row_offset, col_offset}
    vector<vector<pair<int, int>>> directions = {
        {}, // Padding for 0-index
        {{0, -1}, {0, 1}},  // 1: left, right
        {{-1, 0}, {1, 0}},  // 2: upper, lower
        {{0, -1}, {1, 0}},  // 3: left, lower
        {{0, 1}, {1, 0}},   // 4: right, lower
        {{0, -1}, {-1, 0}}, // 5: left, upper
        {{0, 1}, {-1, 0}}   // 6: right, upper
    };

public:
    bool hasValidPath(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        
        queue<pair<int, int>> q;
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        
        q.push({0, 0});
        visited[0][0] = true;
        
        while (!q.empty()) {
            auto [r, c] = q.front();
            q.pop();
            
            if (r == m - 1 && c == n - 1) return true;
            
            int streetType = grid[r][c];
            for (auto& dir : directions[streetType]) {
                int nr = r + dir.first;
                int nc = c + dir.second;
                
                // Check bounds and if visited
                if (nr >= 0 && nr < m && nc >= 0 && nc < n && !visited[nr][nc]) {
                    int nextStreetType = grid[nr][nc];
                    bool backConnect = false;
                    
                    // The neighbor must have a connection back to the current cell
                    // i.e., it must contain the direction (-dir.first, -dir.second)
                    for (auto& backDir : directions[nextStreetType]) {
                        if (backDir.first == -dir.first && backDir.second == -dir.second) {
                            backConnect = true;
                            break;
                        }
                    }
                    
                    if (backConnect) {
                        visited[nr][nc] = true;
                        q.push({nr, nc});
                    }
                }
            }
        }
        
        return false;
    }
};

File: 1391-check-if-there-is-a-valid-path-in-a-grid/solution.py
from collections import deque

class Solution:
    def hasValidPath(self, grid: list[list[int]]) -> bool:
        m, n = len(grid), len(grid[0])
        
        # Mapping street types to their connectivity offsets (dr, dc)
        # 1: Left-Right, 2: Up-Down, 3: Left-Down, 4: Right-Down, 5: Left-Up, 6: Right-Up
        directions = {
            1: [(0, -1), (0, 1)],
            2: [(-1, 0), (1, 0)],
            3: [(0, -1), (1, 0)],
            4: [(0, 1), (1, 0)],
            5: [(0, -1), (-1, 0)],
            6: [(0, 1), (-1, 0)]
        }
        
        queue = deque([(0, 0)])
        visited = set([(0, 0)])
        
        while queue:
            r, c = queue.popleft()
            
            if r == m - 1 and c == n - 1:
                return True
            
            # Explore valid moves from current street type
            for dr, dc in directions[grid[r][c]]:
                nr, nc = r + dr, c + dc
                
                # Check bounds and visited status
                if 0 <= nr < m and 0 <= nc < n and (nr, nc) not in visited:
                    # Check if the next cell connects back to the current cell
                    # Next cell must have an exit pointing in the opposite direction (-dr, -dc)
                    can_connect_back = False
                    for bdr, bdc in directions[grid[nr][nc]]:
                        if bdr == -dr and bdc == -dc:
                            can_connect_back = True
                            break
                    
                    if can_connect_back:
                        visited.add((nr, nc))
                        queue.append((nr, nc))
                        
        return False